### DISCUSSION TITLE
New rpnt (prev. sed) scriptlet

### DISCUSSION THREAD

**Discussion by JobcenterTycoon** [uBlockOrigin/uAssets#18185] [Q&A]
I saw a new commit here https://github.com/gorhill/uBlock/commit/41876336db48292de06707adfa5e97dab74297d2 so i tested the new scriptlet.

But it doesn’t work always on my end. When i add `example.com##+js(sed, #text, Domain)` as a exampe it works when i open the page for the first time but after a reload the scriptlet doesn’t kick in.

Do i did a mistake? Or is this a bug? Tested with Firefox and uBlock Origin 1.49.3b17

**Comment by mapx-** [comment]
It was added a `sed` filter for `ghacks.net` => working on my end in FF
`ghacks.net##+js(sed, #text, /^Advertisement$/)`

- update your lists
- go to https://www.ghacks.net/2023/05/22/twitter-old-tweets-bug/

Do you still see `Advertisement` text  reloading the page ?

**Reply by gorhill** [reply]
The scriptlet acts on mutation events, if the scriptlet is injected such that most of the DOM is already built, it won't act on whatever is already built, as it needs mutation events.

**Reply by JobcenterTycoon** [reply]
This is the point, it works perfectly on `ghacks.net` but the scriptlet fails on other sites like `example.com` this is why im confused. A few times it fails completely even on the first load.

**Reply by mapx-** [reply]
@JobcenterTycoon repeat the test in chrome => working fine on my end for `example.com`

**Reply by JobcenterTycoon** [reply]
Test cases

`ghacks.net`: works on both browsers without issues.
`example.com`: only works on chrome, in firefox only on the first load.
`t-online.de`: Only works on chrome, firefox failing completely (even on the first load).

Example for t-online: `https://www.t-online.de/nachrichten/ausland/usa/id_100179936/tim-scott-dieser-mann-koennte-donald-trump-die-wiederwahl-sichern.html` -> `t-online.de##+js(sed, #text, sichern)` (should remove the last word from the title)

**Reply by gorhill** [reply]
I could reproduce the simple case with `example.com`. The cause is as explained. I just added code to process the current state of the DOM when the scriptlet starts, and this fixes the issue. Ideally, the scriptlet starts when there is just the root `html` node in the DOM such that nodes added afterward will be processed through the mutation observer. But there is no guarantee of this happening, so I will add code to process the already existing nodes. The thing though is that in might not be possible to rewrite JS for script tags that are already in the DOM at the time the scriptlet become live -- nothing can be done for this. It will otherwise work properly for substituting text in other cases.

**Reply by mapx-** [reply]
FF is slower maybe due to  the reason you explained here ?
https://github.com/gorhill/uBlock/commit/4863f3093ee1c1c9edaaad2e9acbd13a1691b599

**Reply by gorhill** [reply]
I instrumented the scriptlet, and when I open either `example.com` or `t-online.de` in a new tab, the scriptlet sees only one node present in the DOM, the expected `html` element. However when I force a reload of the tabs, I get 31 already existing nodes for `example.com`, and 2180 nodes for `t-online.de`, essentially meaning the scriplet seems to have been injected after the page reached `interactive` state in Firefox when forcing a refresh.

**Reply by gorhill** [reply]
> due to the reason you explained here

That code has changed since then, uBO tries to inject as soon as possible (including at that point) regardless of the platform. I am wondering though if it could be because `blob:` is used to inject scriptlet in Firefox .

**Reply by mapx-** [reply]
you mean https://github.com/gorhill/uBlock/commit/0971025b21155f07955ed703143c3324c732cf03

**Reply by MasterKia** [reply]
> The thing though is that in might not be possible to rewrite JS for script tags that are already in the DOM at the time the scriptlet become live

Maybe using HTML filtering on Firefox?

**Reply by gorhill** [reply]
Chromium is quite reliable though, for the same scenario as above, I get "3 nodes present before installing mutation observer" in both cases when forcing a page reload. But there is now an issue: by processing already installed nodes, the scriptlet ends up processing itself...

**Reply by uBlock-user** [reply]
> the scriptlet ends up processing itself...

Because on chromium it's inserted as an inline-script rather than a blob url ?

**Reply by gorhill** [reply]
> wondering though if it could be because `blob:` is used to inject scriptlet in Firefox

The new `sed.js` scriptlet does not have to execute in the _main_ (page) world, since it only deals with DOM nodes (there is no need to trap properties) and as such it can be executed in the _isolated_ (extension's content script) world. This requires lot of changes to support the ability to execute a scriptlet in either world, and now I have the `sed.js` scriptlet injected as a content script rather than a page script, and as a result, Firefox's behavior now seems as good as Chromium. This also simplify a bit things since now the scriptlet does not have to worry about processing itself -- however we need to find out if it can end up processing uBO's scriptlets injected in _main_ world, though I expect not.

I had to adjust the scriptlet a bit since it turns out that in Firefox it is injected earlier than Chromium: with Chromium, there is an `html` element in the DOM at injection time. No so with Firefox, there is nothing in the DOM, so the scriptlet has to be ready for this. So this tells content scripts are also injected reliably in Firefox, the issue is the `blob:`-based injection, it causes the scriptlets in _main_ world to be executed much later. However without `blob:` we have potential security violations in Firefox because of page's CSP.

**Reply by stephenhawk8054** [reply]
Should we ask Rob Wu on slack about Firefox' policy? Better than guessing I think.

**Reply by gorhill** [reply]
Let's not raise the issue, if AdGuard's `replace` is fine, uBO's `sed` should be fine.

**Reply by uBlock-user** [reply]
@stephenhawk8054 ghacks filter no longer works because admin added **space** before and after the word, so change the filter to `ghacks.net##+js(rmnt, #text, /Advertisement/)`

**Comment by gothic-bum** [comment]
Is `replacement` limited to being a string or can it be a regex (which doesn't seem to work in my testing on Firefox)?

If it is a string, how could a comma `,` be included in the replacement text since it is also used as a delimiter?

**Reply by mapx-** [reply]
use `\,`

example:
`example.com##+js(sed, #text, Domain, Hello\,You)`

**Comment by JobcenterTycoon** [comment]
I can confirm it works fine on Firefox now on all 3 tested domains. Also the filter i added for the video ads now working fine on Firefox. Thanks.

**Reply by gorhill** [reply]
Looking at the commit and how the issue was fixed, isn't the case that another solution would have been to have the ability to use `set-constant` with any value, such as `+js(set, props.pageProps.page.thirdParty.homad, disabled)`? Just wondering if such trusted scriptlet should be introduced next.

**Reply by JobcenterTycoon** [reply]
Would be useful for external scripts.

**Reply by MasterKia** [reply]
@gorhill Another use-case:
`photopea.com##+js(trusted-set, window.innerWidth, document.documentElement.clientWidth + 320)`

Related rule in AdGuard Base (to counter ad reinjection placeholder covering 1/4 of the page):
`photopea.com#%#window.innerWidth = document.documentElement.clientWidth + 320;`

https://github.com/AdguardTeam/AdguardFilters/commit/816e810c2ae09c2fa10b5acd0928d99ec610619b

**Reply by gorhill** [reply]
> `window.innerWidth = document.documentElement.clientWidth + 320;`

I don't see how this can work, `window.innerWidth` is a read-only property: <https://developer.mozilla.org/en-US/docs/Web/API/Window/innerWidth>.

**Reply by Yuki2718** [reply]
IIRC https://github.com/uBlockOrigin/uAssets/issues/13369 could have been fixed without exception if trusted-set-constant can difine arbitrary function and was avaialble at the time https://github.com/AdguardTeam/AdguardFilters/commit/66be8cdee82636531468fb768629af75d1e4f395 , but will it be only for value actually?

**Reply by MasterKia** [reply]
> I don't see how this can work

I don't know, but it does work the last time I checked.

**Reply by uBlock-user** [reply]
Executing `window.innerWidth = document.documentElement.clientWidth + 320;` in the console works and the value changes despite being a read-only prop.

**Reply by gorhill** [reply]
> if trusted-set-constant can difine arbitrary function

That would be a clear case of remote code execution, can't have this. A `trusted-set` scriptlet would accept a JSON object with which to set the value, so that it can be any string, number, array with arbitrary content, or object with arbitrary content. For example:

    example.com##+js(trusted-set, prop, { value: 100000 })
    example.com##+js(trusted-set, prop, { value: "yes" })
    example.com##+js(trusted-set, prop, { value: [ "one", "two", 3 ]})
    example.com##+js(trusted-set, prop, { value: { url: "about:blank" }})

**Reply by JobcenterTycoon** [reply]
uBlock Origin can change the size too. When i add `rapidtables.com##+js(set, window.innerWidth, 100)` and open `https://www.rapidtables.com/web/tools/window-size.html` it displays 100 instead of the real size. (on firefox only after clicking "refresh")

**Reply by MasterKia** [reply]
Sure but we need to change it according to `document.documentElement.clientWidth`.

**Reply by gorhill** [reply]
I've added `trusted-set` with <https://github.com/gorhill/uBlock/commit/19cdd50a1cd0ef984f14673cb6e0022cbe2d1b3a>. If the value is meant to be used as a string, no need to use the JSON form, for all other type of value, the JSON form (shown above) is required. So the case:

    example.com##+js(trusted-set, prop, { value: "yes" })

Could be:

    example.com##+js(trusted-set, prop, yes)

**Reply by iam-py-test** [reply]
Can _my filters_ be treated as a trusted location (at least when filterauthormode is enabled)?
Also, OT but the link in the release is broken (the asset still works)

**Reply by mapx-** [reply]
`my filters` is trusted

**Reply by iam-py-test** [reply]
Oh, weird. I'm on 1.49.3b20 and get yellow underline:
![image](https://github.com/uBlockOrigin/uAssets/assets/84232764/d2ba19f8-d7ae-47ac-9f6c-4ddef6b0763b)
Edit: the logger shows it works, though I still don't see prop in devtools. I guess the problem exists between the chair and keyboard

**Reply by mapx-** [reply]
the long form is ok
`example.com##+js(trusted-set-constant, prop, { value: "yes" })`

**Reply by iam-py-test** [reply]
Oh, thanks. No warning but still no prop in devtools. But I guess this is OT

**Reply by gorhill** [reply]
> get yellow underline

~~It's because the scriptlet is not defined, uBO will report it in the logger but nothing happens. If you build uBO yourself, force update of one list, then the scriptlet should be recognized.~~

never mind, the alias is not properly declared, it needs to be `trusted-set.js`, I set it at `trusted-set`. This will be fixed in the next dev build, for now you can use the non-aliased form, `trusted-set-constant` -- and upgrade to 49rc0, 49b20 was broken.

**Reply by iam-py-test** [reply]
Oh, thanks. I don't see that on the releases page, but I can wait until it's available. Thanks for confirming that was the cause

**Reply by gorhill** [reply]
I don't release notes for intra dev build issues, just notes relative to the last stable release.

**Reply by iam-py-test** [reply]
No, I couldn't find that entire release. But turns out, it was just user error on my part; it was there all-along. Thanks and sorry for wasting you time with my stupidity.

**Reply by uBlock-user** [reply]
https://github.com/gorhill/uBlock/commit/879b8376693e435e05f08b05a532d5cfdddff3b3

**Reply by krystian3w** [reply]
> https://github.com/uBlockOrigin/uAssets/discussions/18185#discussioncomment-5980805

Rather impossible as long as it is treated as a string instead of simple arithmetic. It probably won't even help to figure out how to do addition on integers in JSON.

**Reply by BlazeFTL** [reply]
[Screenshot_20230720-201859_Firefox Nightly](https://github.com/uBlockOrigin/uAssets/assets/139938337/53060896-777d-4d22-b232-6d63ae90c84b)
For This Case I Need Trusted-set Right ? I Want To Set
document.visibilityState = visible

**Reply by stephenhawk8054** [reply]
> [Screenshot_20230720-201859_Firefox Nightly](https://github.com/uBlockOrigin/uAssets/assets/139938337/53060896-777d-4d22-b232-6d63ae90c84b)
> For This Case I Need Trusted-set Right ? I Want To Set
> document.visibilityState = visible

Does

```adb
##+js(aeld, visibilitychange)
```

work?

**Reply by BlazeFTL** [reply]
Nope it doesn't..for context it was dropgalaxy 15 sec timer which doesn't get counted if browser is minimized anyway thanks for the help🥺

**Reply by uBlock-user** [reply]
@BlazeFTL  `example.com##+js(trusted-set, document.visibilityState, visible)`

**Reply by BlazeFTL** [reply]
@uBlock-user Working thank you bro😄

**Comment by MasterKia** [comment]
@gorhill What does `sed` stand for? Asking for the wiki.

**Reply by mapx-** [reply]
`sed` is a `find & replace` linux command
(sed from `stream editor`)

**Reply by gorhill** [reply]
I am probably abusing the original command name. I will move `sed` as an alias and provide a more meaningful default name, probably `replace-node-text` (which nobody will use).

**Comment by MasterKia** [comment]
https://github.com/uBlockOrigin/uBlock-issues/issues/2531#issuecomment-1512373333
Can this be fixed using `sed`?

On `https://www.mozilla.org/ja/firefox/browsers/mobile/android/`, the download link has `href="https://app.adjust.com/[...]?redirect=https://play.google.com/[...]`

Can the scriptlet remove `https://app.adjust.com/[...]?redirect=` from `href`?

**Reply by gorhill** [reply]
Unlikely since the URL is uriComponentEncode-ed.

**Reply by krystian3w** [reply]
I would debug why on a Japanese PC/macbook/iMac the a href-sanitizer doesn't work for him (It works fine for me on a "Polish PC" in Firefox and Chrome).

And rather weak is his explanation that he needs a scriptlet already in uBo stable.

**Comment by gorhill** [comment]
> Asking for the wiki.

By the way in last dev build, I rename the scriptlet `replace-node-text.js` (that's essentially what it does), with alias `rnt.js`. `sed.js` still exist just for the sake of the filters already created, should wait for dev build rc4 to be available on Chromium before changing the existing filters.

**Comment by peace2000** [comment]
Fyi, `sed` is now `rpnt`:

https://github.com/gorhill/uBlock/commit/2bb446797a12086f2eebc0c8635b671b8b90c477

**Reply by krystian3w** [reply]
So now you have to wait for 1.49.3rc5 or stable 1.49.4+.

The worst thing is that someone already wanted to do a similar operation with a paragraph in "article" ~~and now I can't find it that discussion, as if he got a shadowban from Microsoft~~ (found and Gorhill also pasted PoC filters to rewrite if someone can't start form 0 aftrr read commits documentation).

**Comment by Yuki2718** [comment]
I tried to remove スポンサードリンク texts visible on `https://rank1-media.com/I0000314` but `rank1-media.com##+js(rpnt, .contents_m, スポンサードリンク)` doesn't work. What did I wrong?

**Reply by gorhill** [reply]
Valid node names are what is listed at <https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeName>. A class name is not a node name. The `rpnt`/`rmnt` scriptlets should _not_ be confused with the currently non-existing procedural cosmetic filter `:remove-text()` -- it works completely differently. Its primary purpose is to disable whole or part of inline javascript, I added the ability to select the target node type because why not, but the primary purpose is not cosmetic. What you have in mind is a procedural cosmetic operator such as `:remove-text()` which would be chained to CSS selectors or other cosmetic procedural operators.

**Reply by gorhill** [reply]
~~The name of the node containing the target text is `article`, and since the target text in present in `view-source:`,~~ Errr see me too I got confused... maybe try:

    rank1-media.com##+js(rmnt, #text, スポンサードリンク)

**Reply by Yuki2718** [reply]
Yes, working! Didn't know `#text` can be used here.

**Comment by MasterKia** [comment]
> `condition, pattern`:
A string or regex which must be found in the text content of the node
in order for the substitution to occur.

`example.com`

`example.com##+js(rpnt, #text, Example Domain, Changed)`
=>
The text of `<title>` (page title) and `<h1>` changes.

***
`example.com##+js(rpnt, #text, Example Domain, Changed, condition, Example)`
=>
Only the text of `<title>` (page title) changes.

**Reply by MasterKia** [reply]
But these two both work:
`example.com##+js(rmnt, #text, Example)`
and
`example.com##+js(rmnt, #text, Example, condition, Exa)`

And this one doesn't work (correctly):
`example.com##+js(rmnt, #text, Example, condition, Exa3434)`

**Reply by krystian3w** [reply]
Also these malform DOM tree:

```adb
example.com##+js(rpnt, , Example Domain, Changed)
example.com##+js(rpnt,, Example Domain, Changed)
```

**Comment by Yuki2718** [comment]
It may have already asked, but will it be possible to replace only part of target text?
Current behavior: remove entire text of target node and then insert replacement text
I want: remove only part of text in the target node specified by the first needle (regex) while the rest of the text unchanged. This means rpnt can be used as a lighter `$replace` for documents. It will also be convenient if the needle can be saved into `$1` like [this filter](https://github.com/AdguardTeam/AdguardFilters/blob/master/BaseFilter/sections/replace.txt#L34): `||imgur.com/*-assets/js/main.*.js$replace=/("hasSeenUploadRewardVideo"\,)!1/\$1true/`

**Reply by gorhill** [reply]
Doesn't this work?

    ...##+js(rpnt, script, /("hasSeenUploadRewardVideo"\,)!1/, $1!0)

**Reply by Yuki2718** [reply]
Ah, so saving string is already supported, that's good. But the main point is to replace only part of the text.

**Reply by gorhill** [reply]
> But the main point is to replace only part of the text

Sorry I don't understand what you are saying. `rpnt` plainly use `String.prototype.replace` internally, which allows for what you are asking. Try in console:

    var a = 'blah blah "hasSeenUploadRewardVideo",!1 blah blah';
    var b = a.replace(/("hasSeenUploadRewardVideo",)!1/, '$1!0')

`b` will then be `blah blah "hasSeenUploadRewardVideo",!0 blah blah`. Isn't this what you want?

**Reply by Yuki2718** [reply]
Sorry, it was just my regex was wrong. I didn't escape `,`, which apparently led to an unintended filter.

**Reply by gorhill** [reply]
You can also use quotes if you don't want to mind commas:

    ...##+js(rpnt, script, '/("hasSeenUploadRewardVideo",)!1/', '$1!0')

**Comment by stephenhawk8054** [comment]
I'm curious about this: in Firefox is it possible to replace text in the document by utilizing `webRequest.filterResponseData` ? Similar to HTML filtering but with capabilities of `rpnt` instead of just removing.

Looking at MDN example, I think it's possible but I'm not sure: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/filterResponseData#examples

**Reply by ghost** [reply]
It's possible, but modifying HTML by RegEx is prone to errors, that's why in uBO source text is parsed by [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser/DOMParser) to DOM tree and standard tools then can be used to filter elements.

https://github.com/gorhill/uBlock/commit/a9f68fe02f40472c3e16287f8a67b0abe9421e10

**Reply by stephenhawk8054** [reply]
I see. Thanks for the information.

Based on this: https://github.com/uBlockOrigin/uAssets/discussions/18185#discussioncomment-5972685 , I think replace text using `filterResponseData` would be more reliable than `rpnt` scriptlet?
